src/model/markercollection.js

```diff
@@ -79,37 +78,47 @@ export default class MarkerCollection {
 	 * Creates and adds a {@link ~Marker marker} to the `MarkerCollection` with given name on given
 	 * {@link module:engine/model/range~Range range}.
 	 *
-	 * If `MarkerCollection` already had a marker with given name (or {@link ~Marker marker} was passed) and the range to
-	 * set is different, the marker in collection is removed and then new marker is added. If the range was same, nothing
-	 * happens and `false` is returned.
+	 * If `MarkerCollection` already had a marker with given name (or {@link ~Marker marker} was passed), the marker in
+	 * collection is updated and {module:engine/model/markercollection~MarkerCollection#event:update} event is fired but only
+	 * if there was a change (marker range or {@link ~Marker#managedUsingOperations} flag has changed.
 	 *
 	 * @protected
-	 * @fires module:engine/model/markercollection~MarkerCollection#event:set
-	 * @fires module:engine/model/markercollection~MarkerCollection#event:remove
+	 * @fires module:engine/model/markercollection~MarkerCollection#event:update
 	 * @param {String|module:engine/model/markercollection~Marker} markerOrName Name of marker to set or marker instance to update.
 	 * @param {module:engine/model/range~Range} range Marker range.
 	 * @param {Boolean} managedUsingOperations Specifies whether the marker is managed using operations.
-	 * @returns {module:engine/model/markercollection~Marker} `Marker` instance added to the collection.
+	 * @returns {module:engine/model/markercollection~Marker} Added or updated `Marker` instance.
 	 */
 	_set( markerOrName, range, managedUsingOperations ) {
 		const markerName = markerOrName instanceof Marker ? markerOrName.name : markerOrName;
 		const oldMarker = this._markers.get( markerName );
 
 		if ( oldMarker ) {
 			const oldRange = oldMarker.getRange();
+			let hasChanged = false;
 
-			if ( oldRange.isEqual( range ) && managedUsingOperations === oldMarker.managedUsingOperations ) {
-				return oldMarker;
+			if ( !oldRange.isEqual( range ) ) {
+				oldMarker._attachLiveRange( LiveRange.createFromRange( range ) );
+				hasChanged = true;
 			}
 
-			this._remove( markerName );
+			if ( !!managedUsingOperations !== oldMarker.managedUsingOperations ) {
```

Comment:
I wonder whether this "Boolean casting" is needed here. If `managedUsingOperations` is a required parameter it should always be `true` or `false`. If it is not, maybe it should have a default value. If the value is taken from somewhere else, maybe it should be cast before the call. Lastly, if `oldMarker.managedUsingOperations` is always a `Boolean` I think you don't need to be overly safe.

Was there an error with condition written as `managedUsingOperations != oldMarker.managedUsingOperations`?

Reply:
BTW. if you cast to `Boolean` on condition you probably should also cast when setting `managedUsingOperations` in a line below.

Reply:
I changed this param to optional and set false as default.